rfcs/20220925-bounded-dynamism.md

```diff
@@ -0,0 +1,258 @@
+  // Add padding to avoid OOM access in the following slice op.
+  %pad_value = stablehlo.constant dense<0.0> : tensor<f32>
+  %padded_data = stablehlo.pad %data, %pad_value, low = [0], high = [7], interior = [0]
+    : (tensor<7xf32>, tensor<f32>) -> tensor<14xf32>
+
+  // Extract the largest possible slice starting at the start index.
+  %scalar_start = stablehlo.reshape %start : (tensor<1xi32>) -> tensor<i32>
+  %padded_result = stablehlo.dynamic_slice %padded_data, %scalar_start, sizes = [7]
+    : (tensor<14xf32>, tensor<i32>) -> tensor<7xf32>
+
+  // Remove the extra elements extracted beyond the limit.
+  %slice_size = stablehlo.subtract %limit, %start : tensor<1xi32>
+  %scalar_size = stablehlo.reshape %slice_size : (tensor<1xi32>) -> tensor<i32>
+  %result = stablehlo.set_dimension_size %padded_result, %scalar_size, dim = 0
+    : (tensor<7xf32>, tensor<i32>)
+      -> tensor<?xf32, #stablehlo.type_extensions<bounds = [7]>>
+
+  func.return %result : tensor<?xf32, #stablehlo.type_extensions<bounds = [7]>>
+}
+
+```
+
+Use of dynamic ops over `set_dimension_size` op has various benefits:
+
+* Greatly simplifies the lowering from higher level frameworks to StableHLO as
+  they don't need to compute upper bounds of dynamic dimensions.
+* Makes lowerings to StableHLO hardware agnostic as they don't depend on if the
+  hardware requires unbounded or bounded programs.
+* Reduces the potential confusion in making use of `set_dimension_size` as the
+  users are generally not familiar with this op and also the semantics are not
+  intuitive.
+
+Benefit of the `set_dimension_size` op:
+
+Given that the runtime size argument of `set_dimension_size` op is required to
+be less than or equal to the static size or bound, compiler could separately
+track runtime size of the tensor and keep a buffer of fixed size according to
+the bound. This helps avoid any data movements for the `set_dimension_size` op
+at the cost of extra memory. However, compilers should be making the trade-off
+between copy and additional memory based on the hardware capabilities and not
+the frameworks. It is possible to lower slice op to `set_dimension_size` op
+easily but going in the other direction is tricky. That would require program
+analysis to make sure that the size of the buffer is not increased later on.
+
+## Alternatives Considered
+
+### Not having bounded type and/or set\_dimension\_size op
+
+Given the use of unbounded dynamism, could StableHLO just have a function
+attribute to store the input bounds instead of having bounded type and
+`set_dimension_size` op? This might be possible but this will pose significant
+challenges for existing users generating bounded programs. Current proposal
+allows users to incrementally move to unbounded dynamism for new implementations
+while immediately making use of StableHLO without generating a mix of StableHLO
+and MHLO programs.
```

Comment:
+ In the [RFC](https://github.com/openxla/stablehlo/blob/71d01753bf79a126eb1a8d5ead03d456689b8d8e/rfcs/20220925-bounded-dynamism.md), the last paragraph is missing in this PR. 
+ The same paragraph is hard to follow. Please rephrase if possible.
+ For clarity, I pasted the content of the paragraph here:

> It is true that having the bounded type and set_dimension_size op introduces some complexity but given that the bounds are optional, users that don't care about bounded dynamism don't encounter complexity. All the code complexity is limited to the StableHLO shape functions. These also affects the op specifications but these should be intuitive to users based on the op semantics and making use of StableHLO shape functions should hide that as well.

Reply:
seems like you are looking at an older version. Removed as it was unnecessary.